File: src/dataset/c/rational-numbers_exercism-yorange1/gemini/GeminiFlash_rational_numbers_v4.c
#include "rational_numbers.h"

#include <math.h>

#include <stdlib.h>
#include <stdbool.h>

static uint16_t gcd(uint16_t a, uint16_t b)
{
    while (b != 0) {
        uint16_t temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

rational_t add(rational_t r1, rational_t r2)
{
    rational_t ret;
    ret.numerator = (int32_t)r1.numerator * r2.denominator + (int32_t)r2.numerator * r1.denominator;
    ret.denominator = (int32_t)r1.denominator * r2.denominator;
    return reduce(ret);
}

rational_t subtract(rational_t r1, rational_t r2)
{
    rational_t ret;
    ret.numerator = (int32_t)r1.numerator * r2.denominator - (int32_t)r2.numerator * r1.denominator;
    ret.denominator = (int32_t)r1.denominator * r2.denominator;
    return reduce(ret);
}

rational_t multiply(rational_t r1, rational_t r2)
{
    rational_t ret;
    if (r1.numerator == 0 || r2.numerator == 0)
    {
        ret.numerator = 0;
        ret.denominator = 1;
        return ret;
    }

    ret.numerator = (int32_t)r1.numerator * r2.numerator;
    ret.denominator = (int32_t)r1.denominator * r2.denominator;

    return reduce(ret);
}

rational_t divide(rational_t r1, rational_t r2)
{
    rational_t ret;
    ret.numerator = (int32_t)r1.numerator * r2.denominator;
    ret.denominator = (int32_t)r1.denominator * r2.numerator;
    
    return reduce(ret);
}

rational_t absolute(rational_t r)
{
    rational_t ret;
    ret.numerator = abs(r.numerator);
    ret.denominator = abs(r.denominator);
    return ret;
}

rational_t reduce(rational_t r)
{
    if (r.numerator == 0) {
        rational_t ret = {0, 1};
        return ret;
    }

    int16_t sign_num = (r.numerator < 0) ? -1 : 1;
    int16_t sign_den = (r.denominator < 0) ? -1 : 1;

    uint16_t num = abs(r.numerator);
    uint16_t den = abs(r.denominator);

    uint16_t common = gcd(num, den);

    rational_t ret;
    ret.numerator = sign_num * (r.numerator / common);
    ret.denominator = sign_den * (r.denominator / common);

    if (ret.denominator < 0) {
        ret.denominator *= -1;
        ret.numerator *= -1;
    }

    return ret;
}

rational_t exp_rational(rational_t r, uint16_t n)
{
    rational_t ret;
    ret.numerator = 1;
    ret.denominator = 1;

    for (uint16_t i = 0; i < n; ++i) {
        ret.numerator *= r.numerator;
        ret.denominator *= r.denominator;
    }

    return reduce(ret);
}

float exp_real(uint16_t x, rational_t r)
{
    return pow(pow(x, (double)r.numerator), 1.0 / r.denominator);
}

File: src/dataset/typescript/rational-numbers_exercism-wackerow/gemini/GeminiFlash_rational-numbers_v4.ts
export default class Rational {
  constructor(private a: number, private b: number) {}

  private gcd(a: number, b: number): number {
    a = Math.abs(a);
    b = Math.abs(b);
    while (b) {
      const temp = b;
      b = a % b;
      a = temp;
    }
    return a;
  }

  public reduce(rational: Rational = new Rational(this.a, this.b)): Rational {
    const gcdValue = this.gcd(rational.a, rational.b);
    let reducedNumerator = rational.a / gcdValue;
    let reducedDenominator = rational.b / gcdValue;

    if (reducedDenominator < 0) {
      reducedNumerator = -reducedNumerator;
      reducedDenominator = -reducedDenominator;
    }

    return new Rational(reducedNumerator, reducedDenominator);
  }

  public add(rational: Rational): Rational {
    const numerator = this.a * rational.b + rational.a * this.b;
    const denominator = this.b * rational.b;
    return this.reduce(new Rational(numerator, denominator));
  }

  public sub(rational: Rational): Rational {
    return this.add(new Rational(-rational.a, rational.b));
  }

  public mul(rational: Rational): Rational {
    const numerator = this.a * rational.a;
    const denominator = this.b * rational.b;
    return this.reduce(new Rational(numerator, denominator));
  }

  public div(rational: Rational): Rational {
    if (rational.a === 0) {
      throw "Cannot divide by zero";
    }
    const numerator = this.a * rational.b;
    const denominator = this.b * rational.a;
    return this.reduce(new Rational(numerator, denominator));
  }

  public abs(): Rational {
    return this.reduce(new Rational(Math.abs(this.a), Math.abs(this.b)));
  }

  public exprational(power: number): Rational {
    if (power === 0) {
      return new Rational(1, 1);
    }

    let numerator: number;
    let denominator: number;

    if (power > 0) {
      numerator = this.a ** power;
      denominator = this.b ** power;
    } else {
      const absPower = Math.abs(power);
      numerator = this.b ** absPower;
      denominator = this.a ** absPower;
    }

    return this.reduce(new Rational(numerator, denominator));
  }

  public expreal(base: number): number {
    return base ** (this.a / this.b);
  }
}